src/repr/src/scalar.rs

```diff
@@ -1653,6 +1653,11 @@ pub trait DatumType<'a, E>: Sized {
     fn into_result(self, temp_storage: &'a RowArena) -> Result<Datum<'a>, E>;
 }
 
+/// A new type that wraps a [`Vec`] that is used to differentiate the target [`Datum`] between
+/// Arrays and Lists. The target of this type is Array.
+#[derive(Debug)]
+pub struct ArrayRustType<T>(pub Vec<T>);
```

Comment:
I don't quite understand why this is needed. It's needed because we have functions that return rust types with unassociated ScalarTypes, and we need a way to tell apart arrays from lists?

Reply:
In order to use the `sqlfunc!` macro, the Rust function provided must return a type that implements `DatumType`. `DatumType` is a trait used to convert from a Rust type to a `Datum` and vice versa. So If I wanted to use `sqlfunc!` to create a SQL functions that returns a SQL array, then the rust function would have to return a type `T` that implements `DatumType` such that `DatumType::into_result` returns a `Datum::Array`.

I could have implemented `DataumType` for a `Vec`, such that `DatumType::into_result` returns a `Datum::Array`. However, that means that all SQL functions that use `sqlfunc!` and return a `Vec`, must return a SQL array. There would be no way to use `sqlfunc!` and return a SQL list.